src/mlpack/core/tree/binary_space_tree/rp_tree_mean_split_impl.hpp

```diff
@@ -0,0 +1,167 @@
+/**
+ * @file rp_tree_mean_split_impl.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Implementation of class (RPTreeMeanSplit) to split a binary space partition
+ * tree.
+ */
+#ifndef MLPACK_CORE_TREE_BINARY_SPACE_TREE_RP_TREE_MEAN_SPLIT_IMPL_HPP
+#define MLPACK_CORE_TREE_BINARY_SPACE_TREE_RP_TREE_MEAN_SPLIT_IMPL_HPP
+
+#include "rp_tree_max_split.hpp"
+
+namespace mlpack {
+namespace tree {
+
+template<typename BoundType, typename MatType>
+bool RPTreeMeanSplit<BoundType, MatType>::SplitNode(const BoundType&  bound,
+                                                  MatType& data,
+                                                  const size_t begin,
+                                                  const size_t count,
+                                                  SplitInfo& splitInfo)
+{
+  const size_t maxNumSamples = 100;
+  const size_t numSamples = std::min(maxNumSamples, count);
+  arma::uvec samples;
+
+  // Get no more than numSamples distinct samples.
+  math::ObtainDistinctSamples(begin, begin + count, numSamples, samples);
+
+  // Find the average distance between points.
+  ElemType averageDistanceSq = GetAveragePointDistance(data, samples);
+
+  const ElemType threshold = 10;
+
+  if (bound.Diameter() * bound.Diameter() <= threshold * averageDistanceSq)
+  {
+    // We will perform the median split.
+    splitInfo.meanSplit = false;
+
+    splitInfo.direction.zeros(data.n_rows);
+
+    // Get a random normal vector.
+    GetRandomDirection(splitInfo.direction);
+
+    // Get the median value of the scalar products of the normal and the
+    // sampled points. The node will be split according to this value.
+    if (!GetDotMedian(data, samples, splitInfo.direction, splitInfo.splitVal))
+      return false;
+  }
+  else
+  {
+    // We will perform the mean split.
+    splitInfo.meanSplit = true;
+
+    // Get the median of the distances between the mean point and the sampled
+    // points. The node will be split according to this value.
+    if (!GetMeanMedian(data, samples, splitInfo.mean, splitInfo.splitVal))
+      return false;
+  }
+
+  return true;
+}
+
+template<typename BoundType, typename MatType>
+typename MatType::elem_type RPTreeMeanSplit<BoundType, MatType>::
+GetAveragePointDistance(
+    MatType& data,
+    const arma::uvec& samples)
+{
+  ElemType dist = 0;
+
+  for (size_t i = 0; i < samples.n_elem; i++)
+    for (size_t j = i + 1; j < samples.n_elem; j++)
+      dist += metric::SquaredEuclideanDistance::Evaluate(data.col(samples[i]),
+          data.col(samples[j]));
+
+  dist /= (samples.n_elem * (samples.n_elem - 1) / 2);
+
+  return dist;
+}
+
+template<typename BoundType, typename MatType>
+void RPTreeMeanSplit<BoundType, MatType>::GetRandomDirection(
+    arma::Col<ElemType>& direction)
```

Comment:
Ah actually, the function is already there!  `math::RandVector()`.  Maybe it would be useful to add an overload with a parameter that allows you to select the number of dimensions.  Note that selecting a random vector by uniformly sampling and then normalizing as you've done here actually causes bias, because you want to uniformly sample from the surface of a sphere.  I believe the implementation Nishant wrote many years ago uses the Box-Muller transform:

https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform